specs/phase1/validator.md

```diff
@@ -0,0 +1,520 @@
+
+A validator should create and broadcast the `attestation` to the associated attestation subnet when either (a) the validator has received a valid `BeaconBlock` from the expected beacon block proposer and a valid `ShardBlock` for the expected shard block proposer for the assigned `slot` or (b) one-half of the `slot` has transpired (`SECONDS_PER_SLOT / 2` seconds after the start of `slot`) -- whichever comes _first_.
+
+#### Attestation data
+
+`attestation_data` is constructed in the same manner as Phase 0 but uses `FullAttestationData` with the addition of two fields -- `head_shard_root` and `shard_transition`.
+
+- Let `head_block` be the result of running the fork choice during the assigned slot.
+- Let `head_state` be the state of `head_block` processed through any empty slots up to the assigned slot using `process_slots(state, slot)`.
+- Let `head_shard_block` be the result of running the fork choice on the assigned shard chain during the assigned slot.
+- Let `shard_blocks` be the shard blocks in the chain starting immediately _after_ the most recent crosslink (`head_state.shard_transitions[shard].latest_block_root`) up to the `head_shard_block`.
+
+*Note*: We assume that the fork choice only follows branches with valid `offset_slots` with respect to the most recent beacon state shard transition for the queried shard.
+
+##### Head shard root
+
+Set `attestation_data.head_shard_root = hash_tree_root(head_shard_block)`.
+
+##### Shard transition
+
+Set `shard_transition` to the value returned by `get_shard_transition(head_state, shard, shard_blocks)`.
+
+```python
+def get_shard_state_transition_result(
+    beacon_state: BeaconState,
+    shard: Shard,
+    shard_blocks: Sequence[SignedShardBlock],
+    validate_signature: bool=True,
+) -> Tuple[Sequence[ShardState], Sequence[Root], Sequence[uint64]]:
+    shard_states = []
+    shard_data_roots = []
+    shard_block_lengths = []
+
+    shard_state = beacon_state.shard_states[shard]
+    shard_block_slots = [shard_block.message.slot for shard_block in shard_blocks]
+    for slot in get_offset_slots(beacon_state, shard):
+        if slot in shard_block_slots:
+            shard_block = shard_blocks[shard_block_slots.index(slot)]
+            shard_data_roots.append(hash_tree_root(shard_block.message.body))
+        else:
+            shard_block = SignedShardBlock(message=ShardBlock(slot=slot))
+            shard_data_roots.append(Root())
+        shard_state = get_post_shard_state(beacon_state, shard_state, shard_block.message)
+        shard_states.append(shard_state)
+        shard_block_lengths.append(len(shard_block.message.body))
+
+    return shard_states, shard_data_roots, shard_block_lengths
+```
+
+```python
+def get_shard_transition(beacon_state: BeaconState,
+                         shard: Shard,
+                         shard_blocks: Sequence[SignedShardBlock]) -> ShardTransition:
+    offset_slots = get_offset_slots(beacon_state, shard)
```

Comment:
Note that to call `get_offset_slots` correctly, the given `beacon_state` should be at *the next slot* of `attestation.data.slot` (where the result `ShardTransition` gets included, not `head_state`.

Alternatively, we can use `compute_offset_slots`.
```python
on_time_slot = Slot(state.slot + 1)
compute_offset_slots(get_latest_slot_for_shard(state, shard), on_time_slot)
```

Reply:
Second thought: https://github.com/ethereum/eth2.0-specs/pull/1875
It seems better to pass the `on_time_slot` (`attestation.data.slot + 1`)

Reply:
note that `head_state` is transitioned up to the assigned slot

Reply:
I'm not certain we need to pass in `on_time_slot` here.

Happy to discuss more though

Reply:
EDIT, I see what you mean here. thinking

Reply:
I think I might prefer not passing in `on_time_slot`. The complexity in #1875 seems pretty high.

I'm exploring the idea a bit more now

Reply:
> note that `head_state` is transitioned up to the assigned slot so head_state already is at the `on_time_slot`

Ah, I see. I thought it's not transitioned head and considered the skipped beacon block case. So using `Slot(state.slot + 1)` is enough for v-guide.

p.s. the complexity of #1875 is mainly in `verify_shard_block_message` change. 😄

Reply:
Yeah, I see. Getting this cleaned up, then going back in there to try to understand!